include/rmm/mr/device/hybrid_memory_resource.hpp

```diff
@@ -0,0 +1,155 @@
+    delete large_size_mr_;
+  }
+
+  /**---------------------------------------------------------------------------*
+   * @brief Queries whether the resource supports use of non-null streams for
+   * allocation/deallocation.
+   *
+   * @returns true
+   *---------------------------------------------------------------------------**/
+  bool supports_streams() const noexcept override { return true; }
+
+ private:
+
+  rmm::mr::device_memory_resource& get_allocator(std::size_t bytes) {
+    if (bytes > max_fixed_size_) return *large_size_mr_;
+    else {
+      auto iter = fixed_size_mr_.begin();
+      while (bytes > iter->get()->get_block_size())
+        iter = std::next(iter);
+      if (iter != fixed_size_mr_.end())
+        return *iter->get();
+      else
+        throw std::bad_alloc();
+    }
+  }
+
+  /**---------------------------------------------------------------------------*
+   * @brief Allocates memory of size at least \p bytes.
+   *
+   * The returned pointer will have at minimum 256 byte alignment.
+   *
+   * @throws std::bad_alloc if size > block_size (constructor parameter)
+   *
+   * @param bytes The size of the allocation
+   * @param stream Stream on which to perform allocation
+   * @return void* Pointer to the newly allocated memory
+   *---------------------------------------------------------------------------**/
+  void* do_allocate(std::size_t bytes, cudaStream_t stream) override {
+    if (bytes <= 0) return nullptr;
+    return get_allocator(bytes).allocate(bytes, stream);
+  }
+
+  /**---------------------------------------------------------------------------*
+   * @brief Deallocate memory pointed to by \p p.
+   *
+   * @throws std::bad_alloc if size > block_size (constructor parameter)
+   *
+   * @param p Pointer to be deallocated
+   * @param bytes The size in bytes of the allocation. This must be equal to the
+   * value of `bytes` that was passed to the `allocate` call that returned `p`.
+   * @param stream Stream on which to perform deallocation
+   *---------------------------------------------------------------------------**/
+  void do_deallocate(void* p, std::size_t bytes, cudaStream_t stream) override {
+    get_allocator(bytes).deallocate(p, bytes, stream);
+  }
+
+  /**---------------------------------------------------------------------------*
+   * @brief Get free and available memory for memory resource
+   *
+   * @throws std::runtime_error if we could not get free / total memory
+   *
+   * @param stream the stream being executed on
+   * @return std::pair with available and free memory for resource
+   *---------------------------------------------------------------------------**/
+  std::pair<std::size_t, std::size_t> do_get_mem_info( cudaStream_t stream) const override {
+    return std::make_pair(0, 0);
+  }
+
+  std::size_t max_fixed_size_; // power of 2 maximum fixed size allocator
+
+  sub_memory_resource *large_size_mr_; // allocator for larger than max_fixed_size
```

Comment:
Just make this a non-pointer member. Then you don't have to remember to delete it in the destructor.

Reply:
Better yet, just make it a parameter (templated).

Reply:
I had this as a non-pointer member but it was getting deleted before the fixed-size resources (that use it) called it's deallocate method to free their memory. How do I control the deletion order if it's not a pointer?  And when you say a parameter, parameter to what?

Reply:
```
template <typename Fixed, typename Big>
class hybrid_memory_resource : public device_memory_resource {

hybrid_memory_resource(Fixed * fixed, Big * big)
```

Something like this. Though it's a little different because you have a vector of `fixed` resources, so I'm not quite sure what the right thing to do is without thinking more about it.

Reply:
I don't think we want the user to have to allocate the internal allocators. I think the class template parameters are the right idea though.

Reply:
There's not much use in making it a template if it isn't also a constructor argument. Otherwise how are you going to know how to construct the upstream resources? 

This allows the user to configure what allocators get used for the fixed/big allocations. Having them be templates but not be constructor arguments means that the allocators can't be stateful, which all of the pool allocators are. 

Check out how Thrust does the same thing with their pool: https://github.com/thrust/thrust/blob/master/thrust/mr/disjoint_pool.h#L108

Reply:
OK, I get it. Right now I'm just trying out different strategies and hard-coding them. And I still have an issue with having the user required to create 5 memory resources in order to create this one that they will use.